impl/src/main/java/org/jboss/weld/util/collections/AbstractMultimap.java

```diff
@@ -44,18 +45,33 @@
     protected final Supplier<C> supplier;
 
     private final Map<K, C> map;
+    private final ConcurrentMap<K, C> concurrentMap;
+    private final boolean concurrentMapProvided;
```

Comment:
This flag seems unnecessary, just check whether `concurrentMap` is null or not. To make the code cleaner I would add a private method `getMap()`:
```
private Map<K, C> getMap() {
   return concurrentMap != null ? concurrentMap : map;
}
```
This method could be used in most of the methods, e.g.:
```
  public int size() {
    return getMap().size();
  }
```